### DISCUSSION TITLE
PyInstaller is broken!

### DISCUSSION THREAD

**Discussion by xytsrm** [pyinstaller/pyinstaller#8434] [PyInstaller]
Recently reinstalled PyInstaller and tried to run it on a .py file it had already created a .exe file back i Sep 2023. 
I set the path to the .py as before, but this time it said it couldn't find the file. (See image "File Not Found")
You'll note that it did create the .spec file in the .py folder.

I copied the file to a lower path; after a deluge of warnings it said couldn't create the .exe because of potential virus or unwanted software (See attached image "Virus".

Had no previous issues in September.
It's also worth noting that the .py file runs without issue using Idle.

X.
![File Not Found](https://github.com/pyinstaller/pyinstaller/assets/92377239/f2a1ecbd-3583-4ed2-bebd-6522dee669d8)
![Virus](https://github.com/pyinstaller/pyinstaller/assets/92377239/22b45c3b-7e7b-4126-b09b-b06c2fee7fcb)

**Comment by bwoodsend** [comment]
Report it to your antivirus vendor then. We can't do anything if they randomly decide to attack the application you're trying to build.

**Comment by rokm** [comment]
The first part of the issue seems to be stemming from the fact that on Windows, `os.path.expandvars` replaces `$$` with `$`; as far as I can tell, the behavior is documented [only in code itself](https://github.com/python/cpython/blob/51aefc5bf907ddffaaf083ded0de773adcdf08c8/Lib/ntpath.py#L433). 

So paths that contain two or more `$` together end up mangled here

https://github.com/pyinstaller/pyinstaller/blob/4d0292c0913b05cf0de1595cab9823baf0c7a478/PyInstaller/building/build_main.py#L1047-L1051

https://github.com/pyinstaller/pyinstaller/blob/4d0292c0913b05cf0de1595cab9823baf0c7a478/PyInstaller/compat.py#L545-L550

I doubt this worked in September, though - you probably just did not use a path with double-`$`.

--- 

The second part of issue is caused by your anti-virus program. You can try building the bootloader from source instead of using pre-compiled ones, but that's about it; if that doesn't help, take it up with your anti-virus vendor.

**Reply by xytsrm** [reply]
What concerns me is that the version of pyinstaller that existed on the PC in Sep23 worked perfectly creating the .exe file. The path has not changed nor has the .py changed. The only change was the installation (reinstallation) of pyinstaller. It's has been reported that hackers have been placing malware into commonly used software on GitHub & PyPl.

I'm using Windows Defender; perhaps a signature update has gotten overly aggressive. But, that still doesn't explain pyinstaller failing to find the .py file, but knows to put the .spec file where the .py is located.

**Reply by bwoodsend** [reply]
You're antivirus's databases are updated every few hours. It's perfectly normal for them to have changed their minds since September.

**Reply by rokm** [reply]
> What concerns me is that the version of pyinstaller that existed on the PC in Sep23 worked perfectly creating the .exe file. The path has not changed nor has the .py changed.

What was the version that worked with that path, then? 

As far as I can tell, that environment-variable-expansion codepath has been there *far* longer than last September...

> It's has been reported that hackers have been placing malware into commonly used software on GitHub & PyPl.

And at the end of the day, it is your choice whether to use software from GitHub & PyPI or not. Same as using pre-compiled PyInstaller bootloaders or building them from the source (after vetting the source yourself).

> But, that still doesn't explain pyinstaller failing to find the .py file, but knows to put the .spec file where the .py is located.

It's because .py file path is stored as relative path in the .spec file, and is made absolute during the build using the spec path; and spec path is computed via the codepath that I linked above, so it mangled the `$$All Python` that you had in your path into `$All Python` - which you can also see in your screenshot.

**Reply by xytsrm** [reply]
If it "mangled" the $$All Python now, why didn't it mangle it in September; the path hasn't changed?

BTW: Are you suggesting that I should ignore the virus warning; is the voluminous messaging prior to the virus warning normal?

**Reply by rokm** [reply]
> If it "mangled" the $$All Python now, why didn't it mangle it in September; the path hasn't changed?

I don't know, that's why I asked you to tell us which version supposedly worked with that path in September. 

I've tested with 6.x, 5.13.2, and 4.10, and they all behave the same in this regard; which is not surprising, considering that environment-variable-expansion codepath has been there since practically forever, and same goes for "'$$' is translated into '$'" behavior of  `ntpath` `expandvars` implementation.

> BTW: Are you suggesting that I should ignore the virus warning; is the voluminous messaging prior to the virus warning normal?

The "virus warning" is not a warning; it is an error caused by your anti-virus program blocking the access to executable while it is being put together. So the build process failed, and even if there is an executable, it is an incomplete one.

What do you mean by "voluminous messaging"? The build process is quite verbose, and you should have seen that in September as well. Or are you referring to the length of the exception message as shown in the screenshot?

**Reply by xytsrm** [reply]
By "voluminous messaging" I simply meant did all that messaging appear as normal; I don't remember quite so much when I did the September build.

Regarding the version not sure I would have known the September version as I simply did a pip install. I believe I have retrieved the original folder from a backup, the name of the overall folder was Python311, the current is Python312. How would I tell the version of PyInstaller in the scripts folder of Python311?

**Reply by rokm** [reply]
If you have everything installed in there, you could try looking into `Python311\lib\site-packages`. In addition to `PyInstaller` package directory, there should also be a `PyInstaller-{version}` metadata directory.

**Reply by xytsrm** [reply]
Think I found it pyinstaller-5.13.2.dist-info. BTW: How do I stop Defender from blocking pyinstaller; I tried a restore,but it didn't restore the .exe file. Would putting pyinstaller in the Defender Exclusions stop the blocking?

I just realized Windows Defender only removed the .exe file recently created by pyinstaller. I just deliberately scanned the .exe files created in September with Defender (with it's up to date signatures) and it did not find and problems with them. This suggests that the virus block is not a false positive. Any comments?

**Reply by rokm** [reply]
So it was 5.13.2. But I'm quite sure it behaves the same as 6.x w.r.t. `$$` in the path. You can try it yourself (either by installing 5.13.2 in your current environment, or by restoring your python 3.11 environment from backup). 

One possible way you could have avoided the issue is if you specified full absolute path to the .py file when running pyinstaller, but even that would result in `build` and `dist` directories being created under `...\$My Stuff\$All Python\MQTT` instead of `...\$My Stuff\$$All Python\MQTT`. Unless your current working directory was also somewhere else, without `$$` in its path.

**Reply by xytsrm** [reply]
Defender is Identifying the .exe produced by PyInstaller as SEVERE. However, I created an exception for the .exe folder. Then I scanned the .exe with two other anti-virus programs; SuperAntiSpyware & MalwareBytes. Neither, found any viruses. So I backed-up my PC and I'm going to test the .exe soon.

**Reply by xytsrm** [reply]
So far no issues with the .exe created by PyInstaller. Curiously, when Defender allegedly found a virus it was placed in the Recycle-Bin. I replaced the file with a duplicate in an excluded folder. I manually ran Malwarebytes on that single file; no virus. then I ran a complete scan with Malwarebytes. Again it did not find any problem with the duplicate file, but it did flag the identical file Defender put in the Recycle Bin; I don't understand why it flagged the file in the Recycle Bin. Anyone have a thought?

**Reply by rokm** [reply]
It could be that difference between the files is that the copy from excluded folder has additional metadata stored in alternate datastreams (or conversely, the original copy has some additional metadata that Windows Defender put there, and the copy from excluded directory does not).

See https://github.com/pyinstaller/pyinstaller/issues/8294#issuecomment-1934680714 and try listing the datastreams for both files.

**Reply by xytsrm** [reply]
> the original copy has some additional metadata that Windows Defender put there

This was what I though was happening, but everything I've read about Defender suggests it doesn't modify any of the files it quarantines or removes. Even tried asking Copilot (Microsoft's ChatGPT), Copilot can be very useful, but sometimes it can get confused and give erroneous information  - lol
At least the "virus problem" seems to be a false-positive by Defender. 

Also, you were right about PyInstaller having problems with the double-$ "$$"; Once I removed the "$$" PyInstaller had no problem finding the target .py file. However, this seems to be a problem with PyInstaller that should be corrected, as no other app I use ever had a issue with "$$".

Thanks for your help.
